### ISSUE TITLE
Can I use the {parallel} package with Shiny and {plumber}?

### ISSUE THREAD

**Issue by schloerke** [rstudio/promises#81]
https://www.linkedin.com/feed/update/urn:li:ugcPost:6899412968317280256?commentUrn=urn%3Ali%3Acomment%3A%28ugcPost%3A6899412968317280256%2C6899428857653334016%29
> That's amazing! I want to try this combo as soon as possible! 🤩
It's a little bit tricky to understand how "future" works 🤔 actually, I can't figure out the real differences with "parallel" and "parallelly" packages and why they can't be used for shiny and plumber

**Comment by HenrikBengtsson** [comment]
Correct, the closest we have to `future()` and `value()` in base R is `mcparallel()` and `mccollect()`, especially since globals are automatically taken care of since forked processing is used  However, the  latter does _not_ handle output and conditions.  For example, 

```sh
> library(parallel)
> f <- mcparallel({ print(1:3); warning("boom"); log("a") })
[1] 1 2 3
> v <- mccollect(f)
> v
$`12609`
[1] "Error in log(\"a\") : non-numeric argument to mathematical function\n"
attr(,"class")
[1] "try-error"
attr(,"condition")
<simpleError in log("a"): non-numeric argument to mathematical function>
```

Note how the output is not captured, and the warning is lost, so those needs to be handled separately.  On the upside, `mcparallel()`/`mccollect()` does indeed preserve the error condition, so you can do something like:

```r
> cond <- attr(v[[1]], "condition")
> if (inherits(cond, "error")) stop(cond)
Error in log("a") : non-numeric argument to mathematical function
```

If you want to do something similar with **parallel** PSOCK clusters, more work is needed, and you'll also  **parallel** handle errors internally (e.g. capture it via a `tryCatch()`, because those are not relayed (you're only getting a try-error object, so you've lost the original error).

These are some of the things that **future** encapsulates in a standard, unified manner:

```sh
> library(future)
> f <- future({ print(1:3); log("a") })
> v <- value(f)
[1] 1 2 3
Error in log("a") : non-numeric argument to mathematical function
```

regardless of parallel backend.  Hopefully that helps clarify the role of **future** here.
